scribe_search/loaders/youtube.py

```diff
@@ -1,5 +1,79 @@
-class Loader:
+import re
+from datetime import time
+from typing import Iterable
+
+from youtube_transcript_api import YouTubeTranscriptApi
+from youtube_transcript_api._errors import NoTranscriptFound
+
+from scribe_search.data import Sub
+from scribe_search.exceptions import InvalidSource
+from scribe_search.loaders import LoaderInterface
+
+
+class Loader(LoaderInterface):
 
     @staticmethod
-    def load(source: str):
-        print("LOADED YOUTUBE")
+    def _extract_video_id(url: str) -> str:
+        """
+        Takes the video id out of a youtube link.
+        """
+        patterns = [
+            r"(?:https?://)?(?:www\.)?youtube\.com/watch\?v=([a-zA-Z0-9_-]{11})",
+            r"(?:https?://)?(?:www\.)?youtu\.be/([a-zA-Z0-9_-]{11})",
+            r"(?:https?://)?(?:www\.)?youtube\.com/embed/([a-zA-Z0-9_-]{11})",
+            r"(?:https?://)?(?:www\.)?youtube\.com/v/([a-zA-Z0-9_-]{11})",
+        ]
+
+        for pattern in patterns:
+            match = re.search(pattern, url)
+            if match:
+                return match.group(1)
+
+        if re.match(r"^[a-zA-Z0-9_-]{11}$", url):
+            return url
+
+        raise InvalidSource(f"Could not extract video ID from URL: {url}")
+
+    @staticmethod
+    def _seconds_to_time(seconds: float) -> time:
+        """
+        Converts seconds to time.
+        """
+        hours = int(seconds // 3600)
+        minutes = int((seconds % 3600) // 60)
+        secs = int(seconds % 60)
+        microseconds = int((seconds % 1) * 1000000)
+        return time(hour=hours, minute=minutes, second=secs, microsecond=microseconds)
+
+    @staticmethod
+    def load(sources: list[str]) -> Iterable[Sub]:
+        """
+        Yields the captions from a link.
+        """
+        api = YouTubeTranscriptApi()
+        for url in sources:
+            try:
+                video_id = Loader._extract_video_id(url)
+
+                try:
+                    captions = api.fetch(video_id)
```

Comment:
Instead of processing the transcript separately, we can format it to srt using

```from youtube_transcript_api.formatters import SRTFormatter```

and the load it with the srt loader. 

Also, the video might have more than one transcript, so api.list(video_id) will fetch all of them. Then format each one and load them with srt Loader.